src/client/game.cpp

```diff
@@ -2517,23 +2488,34 @@ void Game::updatePlayerControl(const CameraOrientation &cam)
 	 */
 	if (m_cache_hold_aux1) {
 		control.aux1 = control.aux1 ^ true;
-		keypress_bits ^= ((u32)(1U << 5));
 	}
 #endif
 
-	LocalPlayer *player = client->getEnv().getLocalPlayer();
+	u32 keypress_bits = (
+			( (u32)(control.jump  & 0x1) << 4) |
+			( (u32)(control.aux1  & 0x1) << 5) |
+			( (u32)(control.sneak & 0x1) << 6) |
+			( (u32)(control.dig   & 0x1) << 7) |
+			( (u32)(control.place & 0x1) << 8) |
+			( (u32)(control.zoom  & 0x1) << 9)
+		);
 
-	// autojump if set: simulate "jump" key
-	if (player->getAutojump()) {
-		control.jump = true;
-		keypress_bits |= 1U << 4;
-	}
+	// Legacy fields to ensure mod compatibility
```

Comment:
Uh no, they're not "legacy". Nothing is being deprecated.

Reply:
They're not legacy at the moment, but this PR *intends* to deprecate them. See this original PR comment for more info: https://github.com/minetest/minetest/pull/10031#issuecomment-771085341

> So, here's my new approach to add analog input: To fully deprecate the forward, backward, left, and right binary inputs, and replace them with a polar coordinate system.

Where does the new code rely on these values? I think I don't need them for base functionality, I just provide them everywhere to stay compatible with the existing code base.

Reply:
In `src/script/lua_api/l_object.cpp`, which runs on the server-side and needs the keypress bits to be sent regardless of what the client uses internally.

> this PR intends to deprecate them. See this original PR comment for more info: #10031 (comment)

This makes sense but you're only looking at the client-side.
The server also uses the player controls but the new `movement_{speed,direction}` is not sent over the network so it's not available there. up/down/left/right cannot be deprecated on the server without a replacement.

Reply:
Hmm yea that makes sense. I couldn't really find any place where the server (disregarding mods / games) makes use of this property, but I don't really know the codebase all that well. That's my bad.

You are definitely right that this information should be supplied to the server. I was mainly worried about changing the client-server API as that sounded like a really big (and even more impacting) change -- especially considering my lack of experience with the MT codebase thus far.

Is there any special "process" we need to follow when proposing such an API change?

Reply:
>  was mainly worried about changing the client-server API as that sounded like a really big change.

It's fine to do this in a later PR especially since this PR works in this state.

> Is there any special "process" we need to follow when proposing such an API change?

If I understand correctly you are referring to the network protocol, calling it an "API" is confusing IMO.
There's no process for this, you just make the change and it has to be backwards-compatible either by bumping the protocol version and explicitly adding code that handles the difference in packets __or__ by appending new data to the end of a packet, which old clients/servers will silently ignore.

Reply:
Yes, I'm talking about the protocol, sorry for the confusion. Thanks for the hints.

Reply:
I think appending `movement_speed` and `movement_direction` to the existing package is less invasive. Will try to add the two fields to `networkprotocol.h`

Reply:
I think it may be a good idea to do that in a later PR to not delay this one further.